update HistogramObserver to be scriptable

lyoka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51001 update HistogramObserver to be scriptable**

fix tests in TestQuantizeJitOps

Differential Revision: [D26038759](https://our.internmc.facebook.com/intern/diff/D26038759)

Commit:
update HistogramObserver to be scriptable

fix tests in TestQuantizeJitOps

[ghstack-poisoned]

Commit:
Update on "update HistogramObserver to be scriptable"


fix tests in TestQuantizeJitOps

Differential Revision: [D26038759](https://our.internmc.facebook.com/intern/diff/D26038759)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_workflow_module.py:
```diff
@@ -695,6 +696,37 @@ def test_observer_scriptable(self, qdtype, qscheme):
         loaded = torch.jit.load(buf)
         self.assertTrue(torch.equal(obs.get_tensor_value()[0], loaded.get_tensor_value()[0]))
 
+class TestHistogramObserver(QuantizationTestCase):
```
Wondering if the tests in def test_save_load_state_dict_script(self) already cover this test.

lyoka replied:
that test scripted the model but didn't save to file, and this problem happens at save step.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -697,7 +697,7 @@ class HistogramObserver(_ObserverBase):
     max_val: torch.Tensor
 
-    def __init__(self, bins=2048, upsample_rate=128, dtype=torch.quint8,
+    def __init__(self, bins: int=2048, upsample_rate=128, dtype=torch.quint8,
```
Can we also add dtype annotations for upsample_rate and dtype?

lyoka replied:
added.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -58,8 +58,8 @@ def __init__(self, dtype):
     @abstractmethod
-    def forward(self, x):
-        pass
+    def forward(self, x: torch.Tensor):
```
Curious as to why this change is needed for scriptability. Since we override forward, shouldnt that be sufficient?

lyoka replied:
reverted

Commit:
Update on "update HistogramObserver to be scriptable"


fix tests in TestQuantizeJitOps

Differential Revision: [D26038759](https://our.internmc.facebook.com/intern/diff/D26038759)

[ghstack-poisoned]

Commit:
Update on "update HistogramObserver to be scriptable"


fix tests in TestQuantizeJitOps

Differential Revision: [D26038759](https://our.internmc.facebook.com/intern/diff/D26038759)

[ghstack-poisoned]
